### PR TITLE
Truncate strings when casting to `time` type

### PR DESCRIPTION
**Pull request by MadhurSikko**
This PR modifies the stringToTimespan(string) (Timespan, error) function in go-mysql-server/sql/types where NULL value was being thrown when there was trailing garbage value unlike SQL which would handle which would handle cases like '12:23 a' and '12:34:12abc'.

Related to dolthub/dolt#10000

### CODE REVIEW

File: enginetest/queries/insert_queries.go [hunk 1]
```diff
@@ -680,6 +680,16 @@ var InsertQueries = []WriteQueryTest{
 			sql.NewRow(5, "5"),
 		},
 	},
+	{
+		WriteQuery:          "INSERT INTO mytable (i, s) VALUES (10, ' 12:23 a'), (11, '11:11:11.12.13'), (12, '10:10:');",
```
Comment (angelamayxie) [inline enginetest/queries/insert_queries.go]: This test doesn't make sense. `mytable.s` is a string type column so it doesn't make sense that these strings would get truncated to a time.

You should write a script test in `InsertScripts` where you create a table with a `time` type column and then insert valid and invalid strings, one at a time, checking for errors and warnings. Then show the entire table to make sure the inserted values are correct.

File: enginetest/queries/queries.go [hunk 2]
```diff
@@ -9557,6 +9557,38 @@ FROM mytable;`,
 			{"DECIMAL"},
 		},
 	},
+	{
+		Query:    "SELECT CAST('  12:23 a' AS TIME);",
+		Expected: []sql.Row{{"12:23:00"}},
+	},
```
Comment (angelamayxie) [inline enginetest/queries/queries.go:9901]: Please check for warning

File: enginetest/queries/queries.go [hunk 3]
```diff
@@ -9557,6 +9557,38 @@ FROM mytable;`,
...
+	{
+		Query:    "SELECT CAST('12:23:18 abc' AS TIME);",
```
Comment (angelamayxie) [inline enginetest/queries/queries.go]: This actually doesn't match MySQL behavior.
```
mysql> select cast('12:23:18 abc' as time);
+------------------------------+
| cast('12:23:18 abc' as time) |
+------------------------------+
| NULL                         |
+------------------------------+
1 row in set, 1 warning (0.00 sec)
```

I understand this is odd because `'12:23:18 a'` does truncate.
```
mysql> select cast('12:23:18 a' as time);
+----------------------------+
| cast('12:23:18 a' as time) |
+----------------------------+
| 12:23:18                   |
+----------------------------+
1 row in set, 2 warnings (0.00 sec)
```

Can you please investigate this?

File: enginetest/queries/queries.go [hunk 4]
```diff
@@ -9557,6 +9557,38 @@ FROM mytable;`,
...
+		Expected: []sql.Row{{"12:23:18"}},
+	},
+	{
+		Query:    "SELECT CAST('12:34:56.123.456' AS TIME);",
+		Expected: []sql.Row{{"12:34:56.123"}},
```
Comment (angelamayxie) [inline enginetest/queries/queries.go]: This doesn't match MySQL behavior
```
mysql> SELECT CAST('12:34:56.123.456' AS TIME);
+----------------------------------+
| CAST('12:34:56.123.456' AS TIME) |
+----------------------------------+
| 12:34:56                         |
+----------------------------------+
1 row in set, 1 warning (0.00 sec)
```

File: enginetest/queries/queries.go [hunk 5]
```diff
@@ -9557,6 +9557,38 @@ FROM mytable;`,
...
+	},
+	{
+		Query:    "SELECT CAST(' -12:00:00 abc' AS TIME);",
+		Expected: []sql.Row{{"-12:00:00"}},
```
Comment (angelamayxie) [inline enginetest/queries/queries.go]: This doesn't match MySQL behavior
```
mysql> SELECT CAST(' -12:00:00 abc' AS TIME);
+--------------------------------+
| CAST(' -12:00:00 abc' AS TIME) |
+--------------------------------+
| 00:00:00                       |
+--------------------------------+
1 row in set, 2 warnings (0.00 sec)
```

Comment (angelamayxie) [reply enginetest/queries/queries.go]: Similar test cases
```
mysql> SELECT CAST('-12:00:00' AS TIME);
+---------------------------+
| CAST('-12:00:00' AS TIME) |
+---------------------------+
| -12:00:00                 |
+---------------------------+
1 row in set (0.00 sec)

mysql> SELECT CAST('-12:00:00asdf' AS TIME);
+-------------------------------+
| CAST('-12:00:00asdf' AS TIME) |
+-------------------------------+
| -12:00:00                     |
+-------------------------------+
1 row in set, 1 warning (0.00 sec)

mysql> SELECT CAST('-12:00:00 asdf' AS TIME);
+--------------------------------+
| CAST('-12:00:00 asdf' AS TIME) |
+--------------------------------+
| 00:00:00                       |
+--------------------------------+
1 row in set, 2 warnings (0.00 sec)
```

File: enginetest/queries/queries.go (hunk 4 above)
Comment (angelamayxie) [reply enginetest/queries/queries.go]: Looks like this is an issue related to dolthub/dolt#10661

Please update the test to the correct value from MySQL and add the `Skip: true` flag along with a comment linking to the related Github issue.

You can add a separate test `SELECT CAST('12:34:56.123456.123' AS TIME(6));` to test similar behavior
```
mysql> SELECT CAST('12:34:56.123456.123' AS TIME(6));
+----------------------------------------+
| CAST('12:34:56.123456.123' AS TIME(6)) |
+----------------------------------------+
| 12:34:56.123456                        |
+----------------------------------------+
1 row in set, 1 warning (0.00 sec)
```

File: sql/types/time.go [hunk 6]
```diff
@@ -323,11 +329,38 @@ func stringToTimespan(s string) (Timespan, error) {
...
+	var dotFound bool
+	for i, r := range s {
+		if !unicode.IsDigit(r) && r != ':' && r != '.' {
+			s = s[0:i]
+			isTruncated = true
+			break
+		}
+	}
+
+	for i, r := range s {
```
Comment (angelamayxie) [inline sql/types/time.go]: I don't think it's necessary to do two passes through the string. Why not do this step during the first pass?

File: sql/types/time.go
Comment (angelamayxie) [inline sql/types/time.go:208]: I think you might actually want to do the string truncation before doing the `stringToTimespan` conversion, instead of part of it. Because you likely want to use a truncated string with `strconv.ParseFloat` and `strconv.ParseInt` (please add tests that touch upon those code paths).

File: sql/types/time.go [hunk 7]
```diff
@@ -323,11 +329,38 @@ func stringToTimespan(s string) (Timespan, error) {
...
+		if r == '.' {
+			if dotFound {
```
Comment (angelamayxie) [inline sql/types/time.go]: `dotFound` is never set to true. So none of this actually ever happens.

File: enginetest/queries/queries.go [hunk 8]
```diff
@@ -9557,6 +9557,38 @@ FROM mytable;`,
...
+	{
+		Query:    "SELECT CAST('00:00:00 garbage' AS TIME);",
+		Expected: []sql.Row{{"00:00:00"}},
+	},
+	{
+		Query:    "SELECT CAST('12:23:' AS TIME);",
+		Expected: []sql.Row{{"12:23:00"}},
+	},
+	{
+		Query:    "SELECT CAST('hello' AS TIME);",
```
Comment (angelamayxie) [inline enginetest/queries/queries.go]: Please check for warnings.

File: enginetest/queries/queries.go [hunk 9]
```diff
@@ -9557,6 +9557,38 @@ FROM mytable;`,
...
+		Expected: []sql.Row{{nil}},
+	},
+	{
+		Query:    "SELECT CAST('   ' AS TIME);",
```
Comment (angelamayxie) [inline enginetest/queries/queries.go]: Please check for warnings.

File: enginetest/queries/queries.go [hunk 10]
```diff
@@ -9557,6 +9557,38 @@ FROM mytable;`,
...
+	{
+		Query:    "SELECT CAST(' -12:00:00 abc' AS TIME);",
+		Expected: []sql.Row{{"-12:00:00"}},
+	},
+	{
+		Query:    "SELECT CAST('00:00:00 garbage' AS TIME);",
+		Expected: []sql.Row{{"00:00:00"}},
+	},
+	{
+		Query:    "SELECT CAST('12:23:' AS TIME);",
```
Comment (angelamayxie) [inline enginetest/queries/queries.go]: Please check for warnings.

File: enginetest/queries/queries.go [hunk 11]
```diff
@@ -9557,6 +9557,38 @@ FROM mytable;`,
...
+	{
+		Query:    "SELECT CAST('12:34:56.123.456' AS TIME);",
+		Expected: []sql.Row{{"12:34:56.123"}},
+	},
+	{
+		Query:    "SELECT CAST(' -12:00:00 abc' AS TIME);",
+		Expected: []sql.Row{{"-12:00:00"}},
+	},
+	{
+		Query:    "SELECT CAST('00:00:00 garbage' AS TIME);",
```
Comment (angelamayxie) [inline enginetest/queries/queries.go]: Please check for warnings.

File: enginetest/queries/queries.go (hunk 3 above)
Comment (MadhurSikko) [reply enginetest/queries/queries.go]: I am not able to figure out the behaviour where it seems that MySQL after hh:mm:ss will have two cases,
1) hh:mm:ss(garbage values): In this case MySQL seems to just remove the garbage values
2) hh:mm:ss (followed by one space and one/two characters): For some reason MySQL seems to be accepting this and anything other than this seems to be giving NULL
Examples:
```
>> select cast('12:23:18.132 aa' as time);
+---------------------------------+
| cast('12:23:18.132 aa' as time) |
+---------------------------------+
| 12:23:18                        |
+---------------------------------+
1 row in set, 2 warnings (0.00 sec)

>> select cast('12:23:18 aam' as time);
+------------------------------+
| cast('12:23:18 aam' as time) |
+------------------------------+
| NULL                         |
+------------------------------+
1 row in set, 1 warning (0.00 sec)

>>  select cast('12:23:18 aaa' as time);show warnings;
+------------------------------+
| cast('12:23:18 aaa' as time) |
+------------------------------+
| NULL                         |
+------------------------------+
1 row in set, 1 warning (0.00 sec)

select cast('12:23:18 jh    ' as time);
+---------------------------------+
| cast('12:23:18 jh    ' as time) |
+---------------------------------+
| NULL                            |
+---------------------------------+
1 row in set, 1 warning (0.00 sec)
```
I am guess that MySQL expects am/pm however it's very lenient, I have been unable to find the proper behaviour in the reference manual and wanted to make sure that I have correctly interpreted the actual behaviour of MySQL before implementing it in Dolt.

additionally, seems to be giving 00:00:00 rather than NULL
```
SELECT CAST('12:00:00 garbage' AS TIME);
+----------------------------------+
| CAST('12:00:00 garbage' AS TIME) |
+----------------------------------+
| 00:00:00                         |
+----------------------------------+
1 row in set, 2 warnings (0.00 sec)
```
However in this case it seems to return NULL so I am assuming there is some sort of time limit afterwards it default to NULL otherwise it return 00:00:00?
```
SELECT CAST('12:30:60 garbage' AS TIME);
+----------------------------------+
| CAST('12:30:60 garbage' AS TIME) |
+----------------------------------+
| NULL                             |
+----------------------------------+
1 row in set, 1 warning (0.00 sec)
```

Also just to make sure I've setup MySQL on strict on my local machine.

Comment (angelamayxie) [reply enginetest/queries/queries.go]: Thank you for investigating this. There are a few points to be made here:

1.
Your last test case should be returning `NULL` even without truncation as it's not a valid time to begin with (60 shouldn't be a valid value in the seconds range).
```
mysql> select cast('12:30:60' as time);
+--------------------------+
| cast('12:30:60' as time) |
+--------------------------+
| NULL                     |
+--------------------------+
1 row in set, 1 warning (0.00 sec)

mysql> show warnings;
+---------+------+--------------------------------------------+
| Level   | Code | Message                                    |
+---------+------+--------------------------------------------+
| Warning | 1292 | Truncated incorrect time value: '12:30:60' |
+---------+------+--------------------------------------------+
1 row in set (0.00 sec)
```
It seems like we do currently do this correctly, but we don't have a warning. It seems like MySQL always includes the `Truncated incorrect time value` warning for any invalid or truncated time. So I think instead of returning `ErrConvertingToTimeType` errors in `stringToTimespan`, you can always return `ErrTruncatedIncorrect` errors. Then you wouldn't need to handle both error cases in `convertValue`. 

2
The difference in char length is certainly odd. It also seems like AM vs PM doesn't actually affect the output so I'm not sure why MySQL does that. It might be worth digging into the MySQL source code to figure out why this is happening, but that's a lot to ask. I'm going to add these cases to the original issue. Please update the tests to match MySQL output and add the `Skip: true` flag with a link to the issue for any failing test cases.

3.
`SELECT CAST('12:00:00 garbage' AS TIME);` returning `00:00:00` seems to be similar to `SELECT CAST(' -12:00:00 abc' AS TIME);` returning `00:00:00`. I'll document this in the original issue. Please update the test cases with the MySQL output and add the `Skip: true` flag with a link to the issue.

4.
Regarding SQL mode, the default SQL modes in Dolt are currently `NO_ENGINE_SUBSTITUTION,ONLY_FULL_GROUP_BY,STRICT_TRANS_TABLES`. Enforcing SQL mode behavior is definitely something we need to work on more so I wouldn't worry about it _too_ much. For now, just set your MySQL to those SQL modes.

Since I'm documenting additional cases in the original issue, I'm going to remove the "fixes" tag in your PR description. Feel free to keep working on this PR. I apologize for this being more complicated than it initially seemed to be -- type-conversion issues are notoriously tricky, and the issue shouldn't have been marked as `good first issue` to begin with (I've since removed the label).

**Review by angelamayxie — CHANGES_REQUESTED** [review]

**Comment by zachmu** [comment]
Merged current main and addressed the review feedback: TIME conversion now emits truncation warnings consistently, truncation happens before numeric fallbacks, MySQL-compatible warning and insert coverage was added, and unsupported edge cases are skipped with issue links. Full go test ./... passes.

**Comment by zachmu** [comment]
@jycor is going to incorporate these changes into a PR for this space that he has in flight. He will close this PR when he does.
